src/storage/postgres/src/main/java/org/locationtech/geogig/storage/postgresql/PGCache.java

```diff
@@ -0,0 +1,214 @@
+        Long maxWeightBytes = maxSize.or(defaultCacheSize());
+
+        return build(initialCapacityCount, concurrencyLevel2, maxWeightBytes, expireSeconds);
+    }
+
+    private static PGCache build(Integer initialCapacityCount, Integer concurrencyLevel2,
+            Long maxWeightBytes, Optional<Integer> expireSeconds) {
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder();
+        cacheBuilder = cacheBuilder.maximumWeight(maxWeightBytes);
+        cacheBuilder.weigher(weigher);
+        if (expireSeconds.isPresent()) {
+            long seconds = expireSeconds.get();
+            Preconditions.checkArgument(seconds > -1);
+            cacheBuilder.expireAfterAccess(seconds, TimeUnit.SECONDS);
+        }
+        cacheBuilder.initialCapacity(initialCapacityCount);
+        cacheBuilder.concurrencyLevel(concurrencyLevel2);
+        cacheBuilder.recordStats();
+
+        SizeTracker sizeTracker = new SizeTracker();
+        cacheBuilder.removalListener(sizeTracker);
+
+        Cache<ObjectId, byte[]> byteCache = cacheBuilder.build();
+
+        return new PGCache(byteCache, sizeTracker);
+    }
+
+    @VisibleForTesting
+    public static PGCache build() {
+        Optional<Integer> expireAfterAccessSeconds = Optional.absent();
+        Integer initialCapacityCount = 1_000_000;
+        Integer concurrencyLevel2 = 16;
+        Long maxWeightBytes = defaultCacheSize();
+
+        return build(initialCapacityCount, concurrencyLevel2, maxWeightBytes,
+                expireAfterAccessSeconds);
+    }
+
+    @VisibleForTesting
+    public void setEncoder(ObjectSerializingFactory encoder) {
+        this.encoder = encoder;
+    }
+
+    private static long defaultCacheSize() {
+        final long maxMemory = Runtime.getRuntime().maxMemory();
+        // Use up to 50% of the heap by default
+        return (long) (maxMemory * 0.5);
+    }
+
+    private Cache<ObjectId, byte[]> cache;
+
+    /**
+     * ConcurrentMap view of the cache, used to check for existent through containsKey instead of
+     * cache.getIfPresent() (to avoid getting the value unnecessarily), and putIfAbsent instead of
+     * cache.put(), to avoid replacing objects since RevObject instances are immutable
+     */
+    private Map<ObjectId, byte[]> map;
+
+    private SizeTracker sizeTracker;
+
+    public PGCache(Cache<ObjectId, byte[]> byteCache, SizeTracker sizeTracker) {
+        this.cache = byteCache;
+        this.sizeTracker = sizeTracker;
+        this.map = cache.asMap();
+    }
+
+    public boolean contains(ObjectId id) {
+        return map.containsKey(id);
+    }
+
+    public void invalidateAll() {
+        cache.invalidateAll();
+        cache.cleanUp();
+    }
+
+    public void dispose() {
+        invalidateAll();
+    }
+
+    public void invalidate(ObjectId id) {
+        cache.invalidate(id);
+    }
+
+    public void put(RevObject obj) {
+        byte[] value = encode(obj);
```

Comment:
might want to check to see if the object is in the map first (before doing the encoding work).
(not quite sure of its use-case if this makes a difference)